### PR TITLE
cuda, nvhpc: cuda@12 with gcc@12

### PR DESCRIPTION
**Pull request by olupton**
Move the NVIDIA toolchain from GCC 11 to GCC 12
- This implies moving to CUDA 12, which needs a driver update on the GPU nodes (HELP-16471)
- This allows us to consistently use GCC 12 in our deployment, which should avoid tedious issues with import order when dynamically loading `neuron+gpu` in `python`, etc.

Other cleanup/tweaks:
- Stop hacking the Spack compilers YAML file to cause `%nvhpc` to load both the `gcc` and `nvhpc` modules. This implies that `gcc` in the Spack build environment for an `%nvhpc` package reverts to being the system GCC (4). When building `neuron` we now explicitly set the CUDA host compiler to be the compiler that was used to build the `cuda` package (GCC12 for CUDA 12, GCC11 for CUDA 11.8) so that we still get a modern toolchain.
- Set `NVHPC_CUDA_HOME` in the `cuda` module to suppress a noisy warning from `nvc++`.
- Drop our patched `cuda` recipe and make the environment modifications in the deployment YAML files instead.

TODO:
- [x] Update the CUDA driver on all BB5 GPU nodes.
- [x] Run the NEURON CI one more time to test the driver update rollout ➡️ https://bbpgitlab.epfl.ch/hpc/cellular/nrn/-/pipelines/114919 ✅

### CODE REVIEW

File: bluebrain/deployment/bin/configure-compilers [hunk 1]
```diff
@@ -77,9 +77,9 @@ while read -r line; do
     fi
 
     GCC_DIR=${DEFAULT_GCC_DIR}
-    if [[ ${spec} = "intel@"* ]]; then
-        GCC_DIR=${LEGACY_GCC_DIR}
-    fi
+    #if [[ ${spec} = "intel@"* ]]; then
+    #    GCC_DIR=${LEGACY_GCC_DIR}
+    #fi
```
Comment (matz-e) [inline bluebrain/deployment/bin/configure-compilers]: That one is no more...

File: bluebrain/repo-patches/packages/cuda/package.py [hunk 2]
```diff
@@ -7,3 +7,7 @@ class Cuda(BuiltinCuda):
     def setup_run_environment(self, env):
         super().setup_run_environment(env)
         env.append_path("LD_LIBRARY_PATH", self.prefix.extras.CUPTI.lib64)
+        # Avoid a warning from nvc++ when the nvhpc and cuda modules
+        # are loaded simultaneously: nvc++-Warning-CUDA_HOME has been
+        # deprecated. Please, use NVHPC_CUDA_HOME instead.
+        env.append_path("NVHPC_CUDA_HOME", self.prefix)
```
Comment (matz-e) [inline bluebrain/repo-patches/packages/cuda/package.py]: So... both these things could actually go into the module inspections in `modules.yaml`?

File: bluebrain/deployment/bin/configure-compilers [hunk 3]
```diff
@@ -118,14 +118,14 @@ while read -r line; do
...
-        if [[ -n $GCC_MODULES ]]; then
-            full_module_path="${GCC_MODULES},${full_module_path}"
-        else
-            log "unable to add GCC module(s) to NVHPC definition"
-            exit 1
-        fi
-    fi
+    #if [[ $line = *"nvhpc"* ]]; then
+    #    if [[ -n $GCC_MODULES ]]; then
+    #        full_module_path="${GCC_MODULES},${full_module_path}"
```
Comment (olupton) [inline bluebrain/deployment/bin/configure-compilers]: I'm not totally sure if we still need this logic or not, so I figured we could test it here.

File: bluebrain/repo-patches/packages/cuda/package.py (hunk 2 above)
Comment (olupton) [reply bluebrain/repo-patches/packages/cuda/package.py]: I tried to do this last night -- is it what you had in mind?

File: bluebrain/repo-patches/packages/neuron/package.py [hunk 4]
```diff
@@ -322,14 +322,21 @@ def cmake_enable_option(spec_requiremement):
...
-                )
+                # Instead of assuming that the gcc in $PATH is the right host compiler, take the
+                # compiler used to build the cuda package as the CUDA host compiler.
+                host_compiler_spec = self.spec["cuda"].compiler
+                # Surely this isn't the best way
+                host_compiler_candidates = [
+                    c for c in spack.compilers.all_compilers() if c.spec == host_compiler_spec
+                ]
+                assert len(host_compiler_candidates) == 1
+                host_compiler = host_compiler_candidates[0]
```
Comment (olupton) [inline bluebrain/repo-patches/packages/neuron/package.py:334]: @matz-e this is..ugly, but I didn't see another way.
This is the alternative to hacking the compiler config in `bin/configure-compilers` to add the relevant GCC module to the list that needs to be loaded for NVHPC.

File: bluebrain/deployment/environments/externals.yaml [hunk 5]
```diff
@@ -47,6 +47,12 @@ spack:
           - ilmbase
           - openblas
           - openexr
+        cuda:
+          environment:
+            append_path:
+              LD_LIBRARY_PATH: '{prefix}/extras/CUPTI/lib64'
+            set:
+              NVHPC_CUDA_HOME: '{prefix}'
```
Comment (olupton) [inline bluebrain/deployment/environments/externals.yaml:55]: Alternative to having a BlueBrain patched recipe for `cuda`.

File: bluebrain/deployment/bin/configure-compilers (hunk 3 above)
Comment (olupton) [reply bluebrain/deployment/bin/configure-compilers]: x-ref: https://github.com/BlueBrain/spack/pull/1929/files#r1152867069

**Review by matz-e — DISMISSED** [review]

File: bluebrain/deployment/bin/configure-compilers [hunk 6]
```diff
@@ -20,18 +20,16 @@ fi
...
-        }
-    }")"
-log "...using gcc ${LEGACY_GCC_VERSION} from ${LEGACY_GCC_DIR} for Intel (legacy) and NVHPC with module(s) ${GCC_MODULES}"
+#GCC_MODULES="$(spack -E config get compilers|sed -n "
+#    /^\s*spec: gcc@${LEGACY_GCC_VERSION}/,/^\s*modules:/ {
+#        /^\s*modules:/ {
+#            s/^\s*modules: \[\(.*\)\]\$/\1/;
+#            p;
+#        }
+#    }")"
```
Comment (pramodk) [inline bluebrain/deployment/bin/configure-compilers:32]: `nitpick`: don't want you to just remove this?

File: bluebrain/repo-patches/packages/neuron/package.py (hunk 4 above)
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py:334]: I agree this is ugly - having this logic in spack recipe 😒. Shouldn't this be enforced by the spec constraint itself?

Comment (olupton) [reply bluebrain/repo-patches/packages/neuron/package.py:334]: I didn't see a simpler way.

When we build `neuron%nvhpc` then we need a CUDA compiler (`nvcc`) for the few `.cu` files we have lying around. We get this via the dependency on `cuda`. For better or for worse, installing `cuda%gcc@12` does not hardcode/configure `nvcc` to use `gcc@12` as its host compiler -- unlike `nvc++` -- so when we actually use `nvcc` (i.e. when building `neuron`) we have to tell it what host compiler to use.

When we're building `neuron%nvhpc`, the only C++ compiler that Spack offers us is `nvc++`, and we generally don't use that as the `nvcc` host compiler, so <some magic> is needed to find the path to `g++` 12.

File: bluebrain/deployment/bin/configure-compilers (hunk 6 above)
Comment (olupton) [reply bluebrain/deployment/bin/configure-compilers:32]: IIUC @matz-e preferred to keep this part lying around

File: bluebrain/repo-patches/packages/neuron/package.py (hunk 4 above)
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py:334]: thanks for reminder!

> For better or for worse, installing cuda%gcc@12 does not hardcode/configure nvcc to use gcc@12 as its host compiler -- unlike nvc++ -- so when we actually use nvcc (i.e. when building neuron) we have to tell it what host compiler to use.

this is annoying for packing aspects..but let's revisit this in a separate ticket/PR.

**Review by pramodk — APPROVED** [review]

**Pull request merged**